lib/apiservers/engine/backends/container.go

```diff
@@ -801,7 +801,71 @@ func (c *Container) ContainerUpdate(name string, hostConfig *containertypes.Host
 // timeout, an error is returned. If you want to wait forever, supply
 // a negative duration for the timeout.
 func (c *Container) ContainerWait(name string, timeout time.Duration) (int, error) {
-	return 0, fmt.Errorf("%s does not implement container.ContainerWait", ProductName())
+	defer trace.End(trace.Begin(fmt.Sprintf("name(%s):timeout(%s)", name, timeout)))
+
+	// Look up the container name in the metadata cache to get long ID
+	vc := cache.ContainerCache().GetContainer(name)
+	if vc == nil {
+		return -1, NotFoundError(name)
+	}
+	id := vc.ContainerID
+
+	//retrieve client to portlayer
+	client := c.containerProxy.Client()
+	results, err := client.Containers.ContainerWait(containers.NewContainerWaitParamsWithContext(ctx).WithTimeout(int64(timeout.Seconds())).WithID(id))
+	if err != nil {
+		switch err := err.(type) {
+		case *containers.ContainerWaitNotFound:
+			return -1, NotFoundError(name)
+		case *containers.ContainerWaitInternalServerError:
+			return -1, InternalServerError(err.Payload.Message)
+		default:
+			return -1, InternalServerError(err.Error())
+		}
+	}
+
+	containerInfo := results.Payload
+
+	// call to the dockerStatus function to retrieve the docker friendly exitCode
+	// TODO: once started / finished time are available replace time.Now()
+	exitCode, _ := dockerStatus(int(*containerInfo.ProcessConfig.ExitCode),
+		*containerInfo.ProcessConfig.Status,
+		*containerInfo.ContainerConfig.State,
+		time.Now(), time.Now())
+
+	return exitCode, nil
+}
+
+// dockerStatus will evaluate the container state, exit code and
+// process status to return a docker friendly status
+func dockerStatus(exitCode int, status string, state string, started time.Time, finished time.Time) (int, string) {
+	// TODO: once started / finished are available then github.com/docker/go-units must be imported for use in
+	// the status string
+	switch state {
+	case "Created":
+		status = state
+	case "Running":
+		//TODO: once we have started use commented out status
+		// status = fmt.Sprintf("Up %s", units.HumanDuration(time.Now().UTC().Sub(started)))
+		status = state
+	case "Stopped":
+
+		if strings.Contains(status, "permission denied") {
```

Comment:
status in this context is the started string from the vmx file -- so in the case of a permission / path /file issue it will have error text.  Unfortunately, in those cases the exit code in the vmx is still 0...possibly that can be corrected in the tether /tools...but I understood that there were some challenges with that.

Agree the use of status as a func variable is confusing -- I'll change that and add some additional explanation.